src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs

```diff
@@ -15,6 +17,125 @@ namespace System.Net.Http
 {
     internal static class ConnectHelper
     {
+        public static ValueTask<Stream> ConnectAsync(string host, int port, bool async, CancellationToken cancellationToken)
+        {
+            return async ? ConnectAsync(host, port, cancellationToken) : new ValueTask<Stream>(Connect(host, port, cancellationToken));
+        }
+
+        private static async ValueTask<Stream> ConnectAsync(string host, int port, CancellationToken cancellationToken)
+        {
+            // We use the static Socket.ConnectAsync with a SocketAsyncEventArgs, because this approach is:
+            // 1. Cancellable
+            // 2. Does not create Dual-stack sockets, which are unavailable in certain environments,
+            //    see https://github.com/dotnet/runtime/issues/44686.
+            var saea = new ConnectEventArgs();
+            try
+            {
+                saea.Initialize(cancellationToken);
+
+                // Configure which server to which to connect.
+                saea.RemoteEndPoint = new DnsEndPoint(host, port);
+
+                // Initiate the connection.
+                if (Socket.ConnectAsync(SocketType.Stream, ProtocolType.Tcp, saea))
+                {
+                    // Connect completing asynchronously. Enable it to be canceled and wait for it.
+                    using (cancellationToken.UnsafeRegister(static s => Socket.CancelConnectAsync((SocketAsyncEventArgs)s!), saea))
+                    {
+                        await saea.Builder.Task.ConfigureAwait(false);
+                    }
+                }
+                else if (saea.SocketError != SocketError.Success)
+                {
+                    // Connect completed synchronously but unsuccessfully.
+                    throw new SocketException((int)saea.SocketError);
+                }
+
+                Debug.Assert(saea.SocketError == SocketError.Success, $"Expected Success, got {saea.SocketError}.");
+                Debug.Assert(saea.ConnectSocket != null, "Expected non-null socket");
+
+                // Configure the socket and return a stream for it.
+                Socket socket = saea.ConnectSocket;
+                socket.NoDelay = true;
+                return new NetworkStream(socket, ownsSocket: true);
+            }
+            catch (Exception error) when (!(error is OperationCanceledException))
```

Comment:
Nit:
```suggestion
            catch (Exception error) when (error is not OperationCanceledException)
```
Reads a bit nicer.

Reply:
I prefer to not do refactors in this PR, there are way too many thinks we may want to fix.